jaxrs/WFLY-11943_Statistics_of_REST_endpoints.adoc

```diff
@@ -0,0 +1,95 @@
+=== Related Issues
+
+* https://issues.jboss.org/browse/WFLY-11751[WFLY-11751]
+* https://issues.jboss.org/browse/WFLY-12091[WFLY-12091]
+
+=== Dev Contacts
+
+* mailto:rsearls@redhat.com[Rebecca Searls]
+* mailto:rsigal@redhat.com[Ron Sigal]
+* mailto:asoldano@redhat.com[Alessio Soldano]
+* mailto:ema@redhat.com[Jim Ma]
+* mailto:weli@redhat.com[Wei Nan Li]
+
+=== QE Contacts
+
+* mailto:mkopecky@redhat.com[Marek Kopecky]
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE. 
+// Discuss with QE during the Kickoff state to decide this
+[X] Engineering
+
+[ ] QE
+
+=== Affected Projects or Components
+
+* WildFly
+* RESTEasy
+
+=== Other Interested Projects
+
+* HAL
+* CLI
+
+== Requirements
+
+* Collect endpoint statistics when instructed by Wildfly.
+* Report accumulated statistics when requested by Wildfly.
+* Reset endpoint statistics when instructed by Wildfly.
+* Stop data collection and clear accumumlated data when instructed by Wildfly.
+* Subsystem attribute `statistics-enabled` will be used to start and stop statistics collection.  Default setting is false.  The default value can be declared to `true` in the Wildfly configuration file, jaxws subsystem element.
+* Subsystem operation `read-resource` will display the statistics for each endpoint.
+* Subsystem operation `reset` will reset collected statistics to zero.
+* Application restart is NOT required when enabling, disabling and resetting statistics collection.
+* Statistics are NOT save on application restart.
+
+=== Hard Requirements
+* Statistics provided
+** `invocation-count` - Total count of calls to endpoint
+** `failed-invocation-count` - Number of endpoint calls that failed
+** `average-execution-time` - The average time in nanosec from the point of endpoint invocation to return.
+** `total-execution-time` - The sum of time in nanosec for all endpoint method invocations.
+
+=== Non-Requirements
+
+* A best effort will be made to report accurate data.  Any lag between report request and data display could provide statistics that are slightly out of date.
+
+* This project is not responsible for the display of the data.  A report request will write the data into Wildfly's reference model for access by other components.
+
+== Test Plan
+One or more test case classes will be added to wildfly/testsuite/integration/basic.
+A test will be provided for each of the following.
+
+* Confirm subsystem attribute, statistics_enabled, can be read and written.
+* Confirm correct invocation-cnt returned on endpoint.
+* Confirm correct failure-cnt returned on endpoint.
+* Confirm statistics provided for each endpoint declared in a resource class.
+* Confirm statistics provided on all endpoints in an app with multiple resource classes.
+* Confirm statistics reset command works.
+
+Tests of this feature in HAL and CLI are not provided.
+
+== Community Documentation
+This feature uses the existing commands and attribute of wildfly metrics
```

Comment:
I created related EAP7-606 jira 3 years ago (before MP Metrics exists). Now, we have MP Metrics.

---

@bstansberry (cc @rsvoboda) This feature is partially duplicate of MP Metrics feature. But AFAIK&&AFAICS:
- MP Metrics needs to use java annotations (like @Metered). Approach of this RFE allows to track statistics without modification of deployment (AFAICS).
- This RFE also adds tracking of `failed-invocation-count` and `total-execution-time` (AFAIK, MP Metrics doesn't allow this).

So WDYT about this RFE? I see [this your comment](https://issues.jboss.org/browse/EAP7-606?focusedCommentId=13733333&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-13733333) on RFE jira that suggests (AFAICS) you are prefering MP Metrics (`microprofile-metrics-smallrye` subsystem) monitoring only, but there is not reaction for answers from @rsearls 

---

@rsearls 
- What do you mean by "wildfly metrics" here? Is it MP Metrics (== `microprofile-metrics-smallrye` subsystem)?
- What is the relationship between this feature (described in this analysis) and MP Metrics?
- I quickly checks [your branch](https://github.com/rsearls/Resteasy/commit/e3e6144b8d5c456fb65c8b72b0a302bfe0712e60#diff-b424c005ca8086734a7b158f672d6604R59) and it seems that this feature won't use MP Metrics, right?
- Do you plan to use Undertow Statistics for this feature (similar like MP Metrics - `/subsystem=undertow/:write-attribute(name=statistics-enabled,value=true)`)?

Can you please mention answers to these questions in the analysis?

Reply:
> MP Metrics - /subsystem=undertow/:write-attribute(name=statistics-enabled,value=true))?
See my comments in the proposal in section Overview.  Attribute, statistics-enabled,
is not unique to MP Metrics. It existed in wfly before version wfly 7.0.0.Final.  This
RFE is following that feature implementation usage.